File: 108-convert-sorted-array-to-binary-search-tree/convert-sorted-array-to-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def convertToBST(self, nums , start, end):
        if start > end:
            return None
        
        mid = start + (end - start) / 2
        parent = TreeNode(nums[mid])
        parent.left = self.convertToBST(nums, start, mid - 1)
        parent.right = self.convertToBST(nums, mid + 1, end)
        
        return parent
    
    def sortedArrayToBST(self, nums):
        """
        :type nums: List[int]
        :rtype: TreeNode
        """
        nums_len = len(nums)
        tree_root = self.convertToBST(nums, 0, nums_len - 1)
        return tree_root
        
        

File: 108-convert-sorted-array-to-binary-search-tree/convert-sorted-array-to-binary-search-tree.go
/*
 * @lc app=leetcode id=108 lang=golang
 *
 * [108] Convert Sorted Array to Binary Search Tree
 *
 * https://leetcode.com/problems/convert-sorted-array-to-binary-search-tree/description/
 *
 * algorithms
 * Easy (50.15%)
 * Total Accepted:    253.3K
 * Total Submissions: 504.6K
 * Testcase Example:  '[-10,-3,0,5,9]'
 *
 * Given an array where elements are sorted in ascending order, convert it to a
 * height balanced BST.
 *
 * For this problem, a height-balanced binary tree is defined as a binary tree
 * in which the depth of the two subtrees of every node never differ by more
 * than 1.
 *
 * Example:
 *
 *
 * Given the sorted array: [-10,-3,0,5,9],
 *
 * One possible answer is: [0,-3,9,-10,null,5], which represents the following
 * height balanced BST:
 *
 * ⁠     0
 * ⁠    / \
 * ⁠  -3   9
 * ⁠  /   /
 * ⁠-10  5
 *
 *
 */
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func toBST(nums []int, low, high int) *TreeNode {
	if low > high {
		return nil
	}

	mid := (low + high) / 2
	node := &TreeNode{
		Val:   nums[mid],
		Left:  toBST(nums, low, mid-1),
		Right: toBST(nums, mid+1, high),
	}
	return node
}

func sortedArrayToBST(nums []int) *TreeNode {
	if len(nums) == 0 {
		return nil
	}
	return toBST(nums, 0, len(nums)-1)
}
